src/main/java/de/tum/in/www1/artemis/service/programming/ProgrammingExerciseParticipationService.java

```diff
@@ -284,4 +287,26 @@ public void resetRepository(VcsRepositoryUrl targetURL, VcsRepositoryUrl sourceU
         gitService.stageAllChanges(targetRepo);
         gitService.commitAndPush(targetRepo, "Reset Exercise", true, null);
     }
+
+    /**
+     * Checks that the result belongs to the default branch of the student participation (in case it has a branch).
+     * For all other cases (template/solution or student participation without a branch) it falls back to check the default branch of the programming exercise.
+     *
+     * @param participation The programming exercise participation in which the submission was made (including a reference to the programming exercise)
+     * @param branchName    The branch received from the CI system.
+     * @throws IllegalStateException Thrown if the result does not belong to the default branch of the exercise.
+     */
+    public void checkCorrectBranchElseThrow(ProgrammingExerciseParticipation participation, String branchName) {
+        String participationDefaultBranch = null;
+        if (participation instanceof ProgrammingExerciseStudentParticipation studentParticipation) {
+            participationDefaultBranch = studentParticipation.getBranch();
```

Comment:
```suggestion
            participationDefaultBranch = versionControlService.get().getOrRetrieveBranchOfStudentParticipation(studentParticipation());
```

Reply:
Maybe we should add a comment to the getter that the getOrRetreive...() should be used instead?